### PR TITLE
Remove uni_fold normalisation and cache url normalisation on annotations

### PR DESCRIPTION
**Pull request by jon-betts**
* This removes the uni_fold normalisation as it was only being applied to ids
 * This adds a caching version of the annotation which keeps normalised results
 * This results in a >5x speed up on my machine (140ms -> 25ms)

### CODE REVIEW

File: tests/h/streamer/filter_test.py [hunk 1]
```diff
@@ -27,20 +27,21 @@ class TestFilterHandler:
...
 
-        ann = factories.Annotation(target_uri=ann_uri)
         assert handler.match(ann) is should_match
 
-    def test_it_matches_id(self, factories):
-        ann_matching = factories.Annotation(target_uri="https://example.com")
-        ann_non_matching = factories.Annotation(target_uri="https://example.net")
+    def test_it_matches_id(self, make_annotation):
+        ann_matching = make_annotation()
+        ann_non_matching = make_annotation()
```
Comment (jon-betts) [inline tests/h/streamer/filter_test.py:44]: The `target_uri` wasn't used in the test and a random one should be provided by the fixture.

File: tests/h/streamer/filter_test.py [hunk 2]
```diff
@@ -27,20 +27,21 @@ class TestFilterHandler:
...
+        ann = make_annotation(target_uri=ann_uri)
+
         query = {
             "match_policy": "include_any",
             "actions": {},
             "clauses": [{"field": "/uri", "operator": "one_of", "value": query_uris}],
         }
         handler = FilterHandler(query)
 
-        ann = factories.Annotation(target_uri=ann_uri)
```
Comment (jon-betts) [inline tests/h/streamer/filter_test.py:38]: Trying to make this a bit more "if, when, then"

File: tests/h/streamer/filter_test.py [hunk 3]
```diff
@@ -67,31 +68,41 @@ def test_it_matches_parent_id(self, factories):
...
+                {
+                    "field": "/id",
+                    "operator": "equals",
+                    "value": "3jgSANNlEeebpLMf36MACw"
+                },
+                {
+                    "field": "/references",
+                    "operator": "one_of",
+                    "value": ["3jgSANNlEeebpLMf36MACw", "3jgSANNlEeebpLMf36MACw"]
+                },
```
Comment (jon-betts) [inline tests/h/streamer/filter_test.py:96]: Add other types of matching to the speed test

File: h/streamer/filter.py [hunk 4]
```diff
@@ -113,21 +118,33 @@ def clause_filter(annotation):
...
-    text = text.lower()
-    text = unicodedata.normalize("NFKD", text)
-    return "".join([c for c in text if not unicodedata.combining(c)])
+    def __init__(self, annotation):
+        self._annotation = annotation
+
+    @property
+    @lru_cache(1)
+    def target_uri(self):
+        return normalize_uri(self._annotation.target_uri)
```
Comment (jon-betts) [inline h/streamer/filter.py:138]: A slight optimisation here too: previously we tested if this was a list, it never is

File: h/streamer/filter.py [hunk 5]
```diff
@@ -113,21 +118,33 @@ def clause_filter(annotation):
...
+
+    @property
+    @lru_cache(1)
+    def references(self):
+        # Ids require no normalisation, so just pass through
+        return self._annotation.references
```
Comment (jon-betts) [inline h/streamer/filter.py:144]: Annoyingly these fields actually have no normalisation applied, but we still need to proxy them

File: h/streamer/filter.py [hunk 6]
```diff
@@ -29,11 +29,25 @@
...
 class FilterHandler:
-    FIELD_GETTERS = {
-        "/uri": attrgetter("target_uri"),
-        "/references": attrgetter("references"),
-        "/id": attrgetter("id"),
+    FIELDS = {
+        # Fields we accept mapped to a getter, and a normalization function
+        "/uri": [attrgetter("target_uri"), _normalize_uri_multi],
+        "/references": [attrgetter("references"), None],
+        "/id": [attrgetter("id"), None],
```
Comment (jon-betts) [inline h/streamer/filter.py:50]: You could use identity lambdas here, but it's about the same speed, and this is less weird

File: h/streamer/filter.py (hunk 5 above)
Comment (robertknight) [reply h/streamer/filter.py:144]: Is there any benefit to the `@lru_cache` decorator here given how simple the method is?

File: h/streamer/filter.py [hunk 7]
```diff
@@ -113,21 +118,33 @@ def clause_filter(annotation):
...
+    @property
+    @lru_cache(1)
+    def references(self):
+        # Ids require no normalisation, so just pass through
+        return self._annotation.references
+
+    @property
+    @lru_cache(1)
+    def id(self):
+        # Ids require no normalisation, so just pass through
```
Comment (robertknight) [inline h/streamer/filter.py:149]: Same question as above, `@lru_cache` is probably not going to be any faster than just executing the method every time.

File: tests/h/streamer/filter_test.py [hunk 8]
```diff
@@ -67,31 +68,41 @@ def test_it_matches_parent_id(self, factories):
...
 
-        ann = factories.Annotation(
+        ann = make_annotation(
             target_uri="https://example.com", references=[other_ann.id]
         )
         assert handler.match(ann) is False
 
-    @pytest.mark.skip(reason="For dev purposes only")
-    def test_speed(self, factories):  # pragma: no cover
+    # @pytest.mark.skip(reason="For dev purposes only")
```
Comment (robertknight) [inline tests/h/streamer/filter_test.py]: It looks like this was left commented-out by mistake.

**Review by robertknight — APPROVED** [review]
LGTM - I tested this locally in the context of the sidebar and single annotation page.

File: h/streamer/filter.py (hunk 5 above)
Comment (jon-betts) [reply h/streamer/filter.py:144]: Yes, a small benefit of only doing the sub-object lookup once. Very minor, but easy to implement

**Pull request merged**